File: mobile/app/src/main/java/vn/lailua/app/logic/Random.kt
package vn.lailua.app.logic

/**
 * Bộ sinh số giả ngẫu nhiên mulberry32 — cùng thuật toán và cùng kết quả với bản web
 * (src/lib/random.ts) để "Đề số N" trên app và trên web là một.
 */
class Rng(seed: Long) {
    private var a: Int = (seed and 0xFFFFFFFFL).toInt()

    fun next(): Double {
        a += 0x6d2b79f5
        var t = a
        t = (t xor (t ushr 15)) * (t or 1)
        t = t xor (t + (t xor (t ushr 7)) * (t or 61))
        return ((t xor (t ushr 14)).toLong() and 0xFFFFFFFFL).toDouble() / 4294967296.0
    }
}

fun <T> List<T>.shuffled(rng: Rng): List<T> {
    val a = toMutableList()
    for (i in a.size - 1 downTo 1) {
        val j = kotlin.math.floor(rng.next() * (i + 1)).toInt()
        val tmp = a[i]; a[i] = a[j]; a[j] = tmp
    }
    return a
}

/** Giống `hash()` của web, kể cả việc JS nhân số thực 64-bit rồi mới ép về 32-bit. */
fun jsHash(n: Long): Long {
    val x = n.toDouble() * 2654435761.0
    val u = toUint32(x)
    val r = u.toInt() xor (u ushr 16).toInt()
    return r.toLong() and 0xFFFFFFFFL
}

private fun toUint32(d: Double): Long {
    val m = kotlin.math.abs(d) % 4294967296.0
    val v = kotlin.math.floor(m).toLong()
    return if (d < 0) (4294967296L - v) and 0xFFFFFFFFL else v
}

/** `key.split("").reduce((a, c) => (a * 31 + c.charCodeAt(0)) % 1_000_000_007, 7)` */
fun jsStringSeed(key: String): Long {
    var a = 7L
    for (c in key) a = (a * 31 + c.code) % 1_000_000_007L
    return a
}

File: ios/Packages/LaiLuaCore/Sources/LaiLuaCore/Random.swift
import Foundation

/// Bộ sinh số giả ngẫu nhiên mulberry32 — cùng thuật toán và cùng kết quả với bản web
/// (src/lib/random.ts) và Android để "Đề số N" trên mọi nền tảng là một.
public struct Rng {
    private var a: UInt32

    public init(seed: UInt64) { a = UInt32(truncatingIfNeeded: seed) }

    public mutating func next() -> Double {
        a = a &+ 0x6d2b79f5
        var t = a
        t = (t ^ (t >> 15)) &* (t | 1)
        t ^= t &+ (t ^ (t >> 7)) &* (t | 61)
        return Double(t ^ (t >> 14)) / 4294967296.0
    }
}

extension Array {
    /// Trộn Fisher–Yates dùng `rng` (giống `shuffle()` của web).
    public func shuffled(using rng: inout Rng) -> [Element] {
        var a = self
        if a.count < 2 { return a }
        for i in stride(from: a.count - 1, through: 1, by: -1) {
            let j = Int((rng.next() * Double(i + 1)).rounded(.down))
            a.swapAt(i, j)
        }
        return a
    }
}

/// Giống `hash()` của web, kể cả việc JS nhân số thực 64-bit rồi mới ép về 32-bit.
public func jsHash(_ n: Int64) -> UInt64 {
    let x = Double(n) * 2654435761.0
    let u = toUint32(x)
    let r = Int32(bitPattern: UInt32(u)) ^ Int32(bitPattern: UInt32(u >> 16))
    return UInt64(UInt32(bitPattern: r))
}

private func toUint32(_ d: Double) -> UInt64 {
    let m = abs(d).truncatingRemainder(dividingBy: 4294967296.0)
    let v = UInt64(m.rounded(.down))
    return d < 0 ? (4294967296 - v) & 0xFFFF_FFFF : v
}

/// `key.split("").reduce((a, c) => (a * 31 + c.charCodeAt(0)) % 1_000_000_007, 7)`
public func jsStringSeed(_ key: String) -> Int64 {
    var a: Int64 = 7
    for unit in key.utf16 { a = (a * 31 + Int64(unit)) % 1_000_000_007 }
    return a
}
